### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by mikebell90**
I had rebase hell, so I just rebuilt. The first commit is the interesting one, the second adds bower_components and hence a lot of whitenoise

Hence I recommend reviewing https://github.com/opentable/request-timeline/pull/35/commits/8643e32fc1fca1eb738b8b5037473120f2e18fe7

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+ENV ROOT=/srv
+ENV SCRIPTS=${ROOT}/scripts
+ENV TIMELINE=${ROOT}/timeline
+ENV NGINX=/etc/nginx
+
+# create directories
+WORKDIR ${ROOT}
+WORKDIR ${SCRIPTS}
+WORKDIR ${TIMELINE}
```
Comment (pennello) [inline Dockerfile]: Not `RUN mkdir ...`?  Seems odd to specify multiple work directories.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+# install nginx
+RUN apt-get update && apt-get install -y nginx
```
Comment (pennello) [inline Dockerfile]: ENV DEBIAN_FRONTEND noninteractive

And, use `--no-install-recommends`.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+COPY css ${TIMELINE}/css
+COPY js ${TIMELINE}/js
+COPY img ${TIMELINE}/img
+COPY index.html ${TIMELINE}
+COPY bower_components ${TIMELINE}/bower_components
+
+# link the site conf (port 8080)
+# remove the default site (port 80)
+# link our site
+RUN ln -s ${TIMELINE} /var/www/timeline && ln -s ${SCRIPTS}/timeline.conf ${NGINX}/sites-available/timeline.conf && ln -s  ${SCRIPTS}/timeline.conf ${NGINX}/sites-enabled/timeline.conf && rm ${NGINX}/sites-enabled/default && rm ${NGINX}/sites-available/default
```
Comment (pennello) [inline Dockerfile]: Can you break some lines here?

Or maybe collapse this stuff into a script that gets added to the container and simply invoked from here?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+# copy files over
+COPY scripts ${SCRIPTS}
+COPY css ${TIMELINE}/css
+COPY js ${TIMELINE}/js
+COPY img ${TIMELINE}/img
+COPY index.html ${TIMELINE}
+COPY bower_components ${TIMELINE}/bower_components
+
+# link the site conf (port 8080)
```
Comment (pennello) [inline Dockerfile]: Won't we want to use `$PORT0`?

File: README.md [hunk 5]
```diff
@@ -5,16 +5,23 @@ This application displays log entries representing HTTP requests
 visually on a timeline.  It depends on services logging according to the
 [common OTLs][1]--specifically, ot-v1, msg-v1, and http-v1.
 
-The service is deployed on the `opengrok` machine, in
-`/var/www/request-timeline`.  It automatically updates to the latest
-`master` code every 10 minutes via a job in
-`/etc/cron.d/request-timeline`.
+On singularity it advertises as `ot-timeline`
```
Comment (pennello) [inline README.md]: I recommend simply `timeline`.

File: build [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh -ex
+
+deploy=$(date +%Y%m%d%H%M%S)
+version=1
+
+echo "building ..."
+docker pull docker.otenv.com/ot-ubuntu:18-latest
```
Comment (pennello) [inline build]: Generate the `Dockerfile`, templating in this `FROM`.  That way, you won't repeat it and get it wrong when one is updated, but not the other.

File: scripts/timeline.conf [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+# applications, such as Drupal or Wordpress. These applications will be made
+# available underneath a path with that package name, such as /drupal8.
+#
+# Please see /usr/share/doc/nginx-doc/examples/ for more detailed examples.
+##
+
+# Default server configuration
+#
+server {
+	listen 8080 default_server;
```
Comment (pennello) [inline scripts/timeline.conf]: Why are we doing this instead of the port from Mesos?

File: scripts/discovery-wrapper [hunk 8]
```diff
@@ -0,0 +1,189 @@
+#!/bin/sh
```
Comment (pennello) [inline scripts/discovery-wrapper]: Why do we even need to launch this into Mesos?  Isn't it a static page?  Why don't we just deploy this in S3?

File: deploy/config/prod-sc/singularity-request.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  "id": "ot-timeline",
+  "requestType": "SERVICE",
+  "owners": [
+    "visibility@opentable.com"
+  ],
+  "instances": 1,
+  "healthcheck": {
+		"startupTimeoutSeconds": 180,
+ 		"startupDelaySeconds": 15,
+ 		"startupIntervalSeconds": 1
```
Comment (pennello) [inline deploy/config/prod-sc/singularity-request.json]: Use consistent indentation.

File: Dockerfile (hunk 1 above)
Comment (mikebell90) [reply Dockerfile]: WORKDIR is documented to create the directory if not exists, and also switch directories. In other words appears to be the equivalent of 
mkdir -p ${dir}
cd ${dir}

File: Dockerfile (hunk 2 above)
Comment (mikebell90) [reply Dockerfile]: done

File: Dockerfile (hunk 3 above)
Comment (mikebell90) [reply Dockerfile]: I was following the recommendations of best practice DockerFile to keep them all to one cached layer. If you do it on multiple lines, you get new layers each time

File: Dockerfile (hunk 4 above)
Comment (mikebell90) [reply Dockerfile]: I guess I could have gone that way. Instead see singularity config...

File: Dockerfile (hunk 1 above)
Comment (pennello) [reply Dockerfile]: Ah.  So instead, you could do `RUN mkdir -p ${SCRIPTS} ${TIMELINE}`.

File: Dockerfile (hunk 3 above)
Comment (pennello) [reply Dockerfile]: Best practices according to whom?  Having many layers has non-trivial overhead.  Readability is important, probably more important than the caching concern here.

File: README.md (hunk 5 above)
Comment (mikebell90) [reply README.md]: Well there are really two items, both of which I went for ot-timeline originally. What to  advertise as discovery, and what to use as a request-Id. I've changed both to 'timeline'

File: build (hunk 6 above)
Comment (mikebell90) [reply build]: I'm sorry, I'm not following. This is the FROM. But that wouldn't update the local cache, would it?

File: scripts/discovery-wrapper (hunk 8 above)
Comment (mikebell90) [reply scripts/discovery-wrapper]: That's what MK requested, I do, and it's simple enough.

File: deploy/config/prod-sc/singularity-request.json (hunk 9 above)
Comment (mikebell90) [reply deploy/config/prod-sc/singularity-request.json]: fixed

File: build (hunk 6 above)
Comment (pennello) [reply build]: cat > Dockerfile <<<EOF
    FROM $tag
    ...

File: scripts/discovery-wrapper (hunk 8 above)
Comment (pennello) [reply scripts/discovery-wrapper]: I see.  Well, I guess we won't be owning this any more, but this is much less simple than just putting some files somewhere.  We have to run a webserver, write up a `singularity.json`, etc.  And maintain the instances when they invariably fail.  Just hosting a static file is much simpler.

File: Dockerfile (hunk 1 above)
Comment (mikebell90) [reply Dockerfile]: which I've done now

File: Dockerfile (hunk 3 above)
Comment (mikebell90) [reply Dockerfile]: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#sort-multi-line-arguments

Comment (mikebell90) [reply Dockerfile]: Anyway I can use line continuation, and that's fine and fixed

**Comment by mikebell90** [comment]
CNAME, FD replaced.

**Pull request merged**